### ISSUE TITLE
JSON_EDITOR causing page to not load

### ISSUE THREAD

**Issue by oboingo** [marcgibbons/django-rest-swagger#573]
After upgrading to DRF 3.5 and django-rest-swagger 2.0.7...

The /docs/ URL for my swagger install loads the page and shows me the docs ednpoint and a few others that do not require being logged in.  I can see my login API, and it has the JSON editor, and it works and lets me login.  Once I use the API to login (or if I click "authorize" and login), I get the same interface and a single line of text stating "Finished Loading Resource Information. Rendering Swagger UI...".  It will sit like that forever I presume.

Watching the Chrome inspector, I can see the request to .../docs/?format=openapi and it returns seemingly valid data.  The javascript console shows:

```
jsoneditor.min.js:7 Uncaught Unknown editor for schema {"description":"","type":"file"}
getEditorClass @ jsoneditor.min.js:7
addObjectProperty @ jsoneditor.min.js:8
(anonymous function) @ jsoneditor.min.js:8
d @ jsoneditor.min.js:7
preBuild @ jsoneditor.min.js:8
f.(anonymous function) @ jsoneditor.min.js:7
(anonymous function) @ jsoneditor.min.js:7
_loadExternalRefs @ jsoneditor.min.js:7
init @ jsoneditor.min.js:7
f @ jsoneditor.min.js:7
render @ swagger-ui.min.js:9
addParameter @ swagger-ui.min.js:9
render @ swagger-ui.min.js:9
addOperation @ swagger-ui.min.js:9
render @ swagger-ui.min.js:9
addResource @ swagger-ui.min.js:8
render @ swagger-ui.min.js:8
render @ swagger-ui.min.js:8
options.success @ swagger-ui.min.js:8
y.buildFromSpec @ swagger-ui.min.js:2
a.finish @ swagger-ui.min.js:3
a.resolve @ swagger-ui.min.js:3
response @ swagger-ui.min.js:2
p @ swagger-ui.min.js:3
t.on.response @ swagger-ui.min.js:3
(anonymous function) @ swagger-ui.min.js:3
f.callback @ swagger-ui.min.js:8
(anonymous function) @ swagger-ui.min.js:8
i.emit @ swagger-ui.min.js:8
n.onreadystatechange @ swagger-ui.min.js:8
```

Looking at the return data from the call I see:
`"photo": {"description": "", "type": "file"}`

...in a model that has a photo on it, declared like:
```
photo = models.ImageField(upload_to=user_photo_directory_path, 
                  null=True, blank=True, 
                  verbose_name=_("photo"))
```

If I set JSON_EDITOR = 'False' (or just comment it out) in the settings, then the Swagger interface loads, but is missing the JSON editor, as expected.

How can I get the editor back?

**Comment by pySilver** [comment]
@marcgibbons any comment on this? it seems that 'file' is not recognized by JSON Editor. I'm wondering if we need it at all.

**Comment by aleksihakli** [comment]
Facing the same issue - JSON editor causes the page not to load, disabling fixes this behaviour when `FileField` OR `ImageField` is used.

Error in the browser console:

    Uncaught Unknown editor for schema {"type":"file","description":""}

**Comment by pySilver** [comment]
@tomchristie this is DRF issue. I've fixed this by overriding `SchemaGenerator.get_serializer_fields` and using:

```python
types_lookup = ClassLookupDict({
    serializers.Field: 'string',
    serializers.IntegerField: 'integer',
    serializers.FloatField: 'number',
    serializers.DecimalField: 'number',
    serializers.BooleanField: 'boolean',
    # serializers.FileField: 'file',  # Cannot be handled by Swagger JSON Editor
    serializers.FileField: 'string',
    serializers.MultipleChoiceField: 'array',
    serializers.ManyRelatedField: 'array',
    serializers.Serializer: 'object',
    serializers.ListSerializer: 'array'
})
```
This looks like wontfix since potentially swagger is not the only one who can use this data.

**Comment by tomchristie** [comment]
> This looks like wontfix

I don't think so - sounds more like we need to be taking more care over how we translate the `type=` in the internal Core API representation into whatever Open API uses.

**Comment by marcgibbons** [comment]
I have not looked into this, but looking at the OpeanAPI spec, `file` parameters can only be used when part of the `form` location. Perhaps the editor is yelling if this is found in the `schema` definition and in the `body`

https://github.com/OAI/OpenAPI-Specification/blob/master//versions/2.0.md#parameterObject

Other possibility is that this is a bug in Swagger UI (or the particular version of Swagger bundled in the project).

Could you try exporting the spec and loading it into a standalone version of SwaggerUI to rule it out?

@tomchristie How are file fields being handled at the moment? We likely need some special handling for endpoints containing these and render the spec so the parameters are in `form` rather than body/schema.

**Comment by tomchristie** [comment]
I wasn't able to replicate this with the latest versions. Don't know if the Swagger UI version has been updated between those, and if it now handles that error case gracefully or not.

The latest version of the Open API codec should resolve this: 1.2.1

That'll ensure that `type="file"` gets coerced to `type="string"` for JSON inputs.

There's also an associated issue that we could make the schema generation more graceful in REST framework for this case too... `get_encoding` could check if there's a `file` field present *and* no nested fields. In that case it could prefer multipart encoding (if available in the parsers)
